README.md
# tinycolor2
> TinyColor is a small library for Flutter color manipulation and conversion

A port of [tinycolor2](https://github.com/bgrins/TinyColor) by [Brian Grinstead](https://github.com/bgrins)

## Getting Started

A `TinyColor` receives a `Color` as parameter.

```dart
import 'package:tinycolor2/tinycolor2.dart';

final TinyColor tinyColor = TinyColor(Colors.green);
```
Now you can also use the package to extend the native `Color` class with all the same features, but simpler. To use extension update, make sure to change envieronment sdk version in pubspec like this: ` sdk: ">=2.6.0 <3.0.0"`

## Constructors

### From a Hex String

The package uses [Pigment](https://pub.dartlang.org/packages/pigment) by [Bregy Malpartida Ramos](https://github.com/bregydoc/) to convert strings to `Color`

```dart
TinyColor.fromString('#FE5567');
```

### From RGB int values

```dart
TinyColor.fromRGB(r: 255, g: 255, b:255);
```

### From `HslColor`

```dart
final HslColor color = HslColor(h: 250, s: 57, l: 30);
TinyColor.fromHSL(color);
```

### From `HSVColor`

```dart
final HSVColor color = HSVColor.fromAHSV(1.0, 250, 57, 30);
TinyColor.fromHSV(color);
```

### From Flutter's `Color`

```dart
final TinyColor tinyColor = Colors.blue.toTinyColor();
```

## Properties

### color

Returns the flutter `Color` after operations 

```dart
final Color color = TinyColor(Colors.white).color;
```

## Methods

### getBrightness

Returns the perceived brightness of a color, from `0` to `255`, as defined by [Web Content Accessibility Guidelines (Version 1.0)](https://www.w3.org/TR/AERT#color-contrast).

```dart
TinyColor.fromString("#ffffff").getBrightness(); // 255
TinyColor.fromString("#000000").getBrightness(); // 0
// or with Color extension
Colors.grey.brightness;  // 127
```

### isLight

Return a boolean indicating whether the color's perceived brightness is light.

```dart
TinyColor.fromString("#ffffff").isLight(); // true
TinyColor.fromString("#000000").isLight(); // false
// or with Color extension
Colors.white.isLight;  // true
```

### isDark

Return a boolean indicating whether the color's perceived brightness is dark.

```dart
TinyColor.fromString("#ffffff").isDark(); // false
TinyColor.fromString("#000000").isDark(); // true
// or with Color extension
Colors.white.isDark;  // false
```

### getLuminance

Return the perceived luminance of a color, a shorthand for flutter `Color.computeLuminance`

```dart
TinyColor.fromString("#ffffff").getLuminance();
// or with Color extension
Colors.white.luminance;
```

## Color Modification

These methods manipulate the current color, and return it for chaining. For instance:

```dart
TinyColor(Colors.red).lighten().desaturate().color;
// or with Color extension
Colors.red.lighten().desaturate();
```

### setAlpha

Sets the alpha value on the current color, from `0` to `100`.

```dart
TinyColor(Colors.red).setAlpha(10);
```

### setOpacity

Sets the opacity value on the current color, from `0.0` to `1.0`.

```dart
TinyColor(Colors.red).setOpacity(0.5);
```

### lighten

`lighten: function(amount = 10) -> TinyColor`. Lighten the color a given amount, from `0` to `100`. Providing `100` will always return white.

```dart
TinyColor(Colors.red).lighten().color;
TinyColor(Colors.red).lighten(100).color;
// or with Color extension
Colors.red.lighten(50);
```

### brighten

`brighten: function(amount = 10) -> TinyColor`. Brighten the color a given amount, from `0` to `100`.

```dart
TinyColor(Colors.black).brighten().color;
// or with Color extension
Colors.black.brighten(50);
```

### darken

`darken: function(amount = 10) -> TinyColor`. Darken the color a given amount, from `0` to `100`. Providing `100` will always return black.

```dart
TinyColor(Colors.red).darken().color;
TinyColor(Colors.red).darken(100).color;
// or with Color extension
Colors.red.darken(50);
```

### tint

Mix the color with pure white, from `0` to `100`. Providing `0` will do nothing, providing `100` will always return white.

```dart
TinyColor(Color.red).tint().color;
TinyColor(Color.red).tint(100).color;
// or with Color extension
Colors.red.tint(50);
```

### shade

Mix the color with pure black, from `0` to `100`. Providing `0` will do nothing, providing `100` will always return black.

```dart
TinyColor(Colors.red).shade().color;
TinyColor(Colors.red).shade(100).color;
// or with Color extension
Colors.red.shade(50);
```

### desaturate

`desaturate: function(amount = 10) -> TinyColor`. Desaturate the color a given amount, from `0` to `100`. Providing `100` will is the same as calling `greyscale`.

```dart
TinyColor(Colors.red).desaturate().color;
TinyColor(Colors.red).desaturate(100).color;
// or with Color extension
Colors.red.desaturate(50);
```

### saturate

`saturate: function(amount = 10) -> TinyColor`. Saturate the color a given amount, from `0` to `100`.

```dart
TinyColor(Colors.red).saturate().color;
// or with Color extension
Colors.red.saturate(50);
```

### greyscale

`greyscale: function() -> TinyColor`. Completely desaturates a color into greyscale. Same as calling `desaturate(100)`.

```dart
TinyColor(Colors.red).greyscale().color;
// or with Color extension
Colors.red.greyscale;
```

### spin

`spin: function(amount = 0) -> TinyColor`. Spin the hue a given amount, from `-360` to `360`. Calling with `0`, `360`, or `-360` will do nothing (since it sets the hue back to what it was before).

```dart
TinyColor(Colors.red).spin(180).color;
// or with Color extension
Colors.red.spin(180);

// spin(0) and spin(360) do nothing
TinyColor(Colors.red).spin(0).color;
TinyColor(Colors.red).spin(360).color;
```

### compliment

`compliment: function() -> TinyColor`. Returns the Complimentary Color for dynamic matching.

```dart
TinyColor(Colors.red).compliment().color;
// or with Color extension
Colors.red.compliment;
```

### mix

`mix: function(toColor, amount = 10) -> TinyColor`. Blends the color with another color a given amount, from `0` to `100`, default `10`.

```dart
TinyColor(Colors.red).mix(TinyColor(Colors.yellow, 20)).color;
// or with Color extension
Colors.red.mix(Colors.yellow, 20);
```

## Common operations

### clone

`clone: function() -> TinyColor`. Instantiate a new `TinyColor` object with the same color. Any changes to the new one won't affect the old one.

```dart
final TinyColor color1 = TinyColor(Colors.red);
final TinyColor color2 = color1.clone();
color2.setAlpha(20);
```

### ==

`==: function(Object) -> bool`. Compares if `[Object]` is the same `TinyColor` object.

```dart
final TinyColor color1 = TinyColor(Colors.blue);
final TinyColor color2 = TinyColor(Colors.yellow);
if (color1 == color2) return "same";
else return "different";
```


CHANGELOG.md
# Changelog

## [2.0.0] - 2021-07-27

### Added
* `.toHex8()` at [PR #17](https://github.com/TinyCommunity/tinycolor2/pull/17)
* `.equals()` and `operator ==` at [PR #16](https://github.com/TinyCommunity/tinycolor2/pull/16)
* Example project at [PR #13](https://github.com/TinyCommunity/tinycolor2/pull/13)

### Changed
* Forked as a community version with new name `tinycolor2`
* Null safety migration w/ dependency upgrade at [PR #24](https://github.com/TinyCommunity/tinycolor2/pull/24)
* Library restructure at [PR #5](https://github.com/TinyCommunity/tinycolor2/pull/5)

### Fixed
* #2, #6, #10 from upstream at [PR #5](https://github.com/TinyCommunity/tinycolor2/pull/5)

## [1.0.3] - 2020-08-27

### Added
* `Color` class extension method for direct usage

## [1.0.2] - 2018-08-17

### Fixed
* Compatibility issues with Dart `2.1` and `pigment` `1.0.3`

## [1.0.1] - 2018-08-04

### Fixed
* `HSLColor` class name duplicate definition

### Changed
* Package description

## [1.0.0] - 2018-08-04

### Added
* Initial port of `tinycolor2`


test/tinycolor_test.dart
import 'dart:ui';

import 'package:flutter_test/flutter_test.dart';
import 'package:tinycolor2/tinycolor2.dart';

void main() {
  test(
    "setAlpha updates alpha value of color",
    () {
      final TinyColor color = TinyColor(const Color(0xFFFFFFFF));
      color.setAlpha(0x00);
      expect(color.color.alpha, 0x00);
    },
  );

  test(
    "setOpacity updates opacity value of color",
    () {
      final TinyColor color = TinyColor(const Color(0xFFFFFFFF).withOpacity(1.0));
      color.setOpacity(0.5);

      // underlying dart implementation converts the opacity value to an
      // int, then back into a double. Thus some precision is loss.
      expect(color.color.opacity, moreOrLessEquals(0.5, epsilon: 1e-2));
    },
  );
}


lib/src/color_extension.dart
import 'package:flutter/painting.dart';
import 'tinycolor.dart';

/// Extends the Color class to allow direct TinyColor manipulation natively
extension TinyColorExtension on Color {
  /// Converts standard Color to TinyColor object
  TinyColor toTinyColor() => TinyColor(this);

  HSVColor toHsv() => TinyColor(this).toHsv();

  HslColor toHsl() => TinyColor(this).toHsl();

  /// Lighten the color a given amount, from 0 to 100. Providing 100 will always return white.
  Color lighten([int amount = 10]) => TinyColor(this).lighten(amount).color;

  /// Brighten the color a given amount, from 0 to 100.
  Color brighten([int amount = 10]) => TinyColor(this).brighten(amount).color;

  /// Darken the color a given amount, from 0 to 100. Providing 100 will always return black.
  Color darken([int amount = 10]) => TinyColor(this).darken(amount).color;

  /// Mix the color with pure white, from 0 to 100. Providing 0 will do nothing, providing 100 will always return white.
  Color tint([int amount = 10]) => TinyColor(this).tint(amount).color;

  /// Mix the color with pure black, from 0 to 100. Providing 0 will do nothing, providing 100 will always return black.
  Color shade([int amount = 10]) => TinyColor(this).shade(amount).color;

  /// Desaturate the color a given amount, from 0 to 100. Providing 100 will is the same as calling greyscale.
  Color desaturate([int amount = 10]) => TinyColor(this).desaturate(amount).color;

  /// Saturate the color a given amount, from 0 to 100.
  Color saturate([int amount = 10]) => TinyColor(this).saturate(amount).color;

  /// Completely desaturates a color into greyscale. Same as calling desaturate(100).
  Color get greyscale => TinyColor(this).greyscale().color;

  /// Spin the hue a given amount, from -360 to 360. Calling with 0, 360, or -360 will do nothing (since it sets the hue back to what it was before).
  Color spin([double amount = 0]) => TinyColor(this).spin(amount).color;

  /// Returns the perceived brightness of a color, from 0-255, as defined by Web Content Accessibility Guidelines (Version 1.0).Returns the perceived brightness of a color, from 0-255, as defined by Web Content Accessibility Guidelines (Version 1.0).
  double get brightness => TinyColor(this).getBrightness();

  /// Return the perceived luminance of a color, a shorthand for flutter Color.computeLuminance
  double get luminance => TinyColor(this).getLuminance();

  /// Return a boolean indicating whether the color's perceived brightness is light.
  bool get isLight => TinyColor(this).isLight();

  /// Return a boolean indicating whether the color's perceived brightness is dark.
  bool get isDark => TinyColor(this).isDark();

  /// Returns the Complimentary Color for dynamic matching
  Color get compliment => TinyColor(this).complement().color;

  /// Blends the color with another color a given amount, from 0 - 100, default 50.
  Color mix(Color toColor, [int amount = 50]) => TinyColor(this).mix(
    input: toColor,
    amount: amount,
  ).color;
}


lib/src/tinycolor.dart
import 'dart:math' as math;
import 'dart:ui' show Color;

import 'package:flutter/painting.dart' show HSVColor;
import 'package:pigment/pigment.dart';

import 'conversion.dart';
import 'hsl_color.dart';
import 'util.dart';

class TinyColor {
  final Color originalColor;
  Color _color;

  TinyColor(this.originalColor) : _color = originalColor.clone();

  factory TinyColor.fromRGB({
    required int r,
    required int g,
    required int b,
    int a = 100,
  }) =>
      TinyColor(Color.fromARGB(a, r, g, b));

  factory TinyColor.fromHSL(HslColor hsl) => TinyColor(hslToColor(hsl));

  factory TinyColor.fromHSV(HSVColor hsv) => TinyColor(hsv.toColor());

  factory TinyColor.fromString(String string) =>
      TinyColor(Pigment.fromString(string));

  bool isDark() => getBrightness() < 128.0;

  bool isLight() => !isDark();

  double getBrightness() =>
      (_color.red * 299 + _color.green * 587 + _color.blue * 114) / 1000;

  double getLuminance() => _color.computeLuminance();

  TinyColor setAlpha(int alpha) {
    _color = _color.withAlpha(alpha);
    return this;
  }

  TinyColor setOpacity(double opacity) {
    _color = _color.withOpacity(opacity);
    return this;
  }

  HSVColor toHsv() => HSVColor.fromColor(_color);

  HslColor toHsl() {
    final hsl = rgbToHsl(
      r: _color.red.toDouble(),
      g: _color.green.toDouble(),
      b: _color.blue.toDouble(),
    );
    return HslColor(
      h: hsl.h * 360,
      s: hsl.s,
      l: hsl.l,
      a: _color.alpha.toDouble(),
    );
  }

  String toHex8() => _color.value.toRadixString(16).padLeft(8, '0');

  TinyColor clone() => TinyColor(_color);

  TinyColor lighten([int amount = 10]) {
    final hsl = toHsl();
    hsl.l += amount / 100;
    hsl.l = clamp01(hsl.l);
    return TinyColor.fromHSL(hsl);
  }

  TinyColor brighten([int amount = 10]) {
    final color = Color.fromARGB(
      _color.alpha,
      math.max(
        0,
        math.min(
          255,
          _color.red - (255 * -(amount / 100)).round(),
        ),
      ),
      math.max(
        0,
        math.min(
          255,
          _color.green - (255 * -(amount / 100)).round(),
        ),
      ),
      math.max(
        0,
        math.min(
          255,
          _color.blue - (255 * -(amount / 100)).round(),
        ),
      ),
    );
    return TinyColor(color);
  }

  TinyColor darken([int amount = 10]) {
    final hsl = toHsl();
    hsl.l -= amount / 100;
    hsl.l = clamp01(hsl.l);
    return TinyColor.fromHSL(hsl);
  }

  TinyColor tint([int amount = 10]) => mix(
        input: const Color.fromRGBO(255, 255, 255, 1.0),
        amount: amount,
      );

  TinyColor shade([int amount = 10]) => mix(
        input: const Color.fromRGBO(0, 0, 0, 1.0),
        amount: amount,
      );

  TinyColor desaturate([int amount = 10]) {
    final hsl = toHsl();
    hsl.s -= amount / 100;
    hsl.s = clamp01(hsl.s);
    return TinyColor.fromHSL(hsl);
  }

  TinyColor saturate([int amount = 10]) {
    final hsl = toHsl();
    hsl.s += amount / 100;
    hsl.s = clamp01(hsl.s);
    return TinyColor.fromHSL(hsl);
  }

  TinyColor greyscale() => desaturate(100);

  TinyColor spin(double amount) {
    final hsl = toHsl();
    final hue = (hsl.h + amount) % 360;
    hsl.h = hue < 0 ? 360 + hue : hue;
    return TinyColor.fromHSL(hsl);
  }

  TinyColor mix({
    required Color input,
    int amount = 50,
  }) {
    final p = amount / 100.0;
    final color = Color.fromARGB(
        ((input.alpha - _color.alpha) * p + _color.alpha).round(),
        ((input.red - _color.red) * p + _color.red).round(),
        ((input.green - _color.green) * p + _color.green).round(),
        ((input.blue - _color.blue) * p + _color.blue).round());
    return TinyColor(color);
  }

  TinyColor complement() {
    final hsl = toHsl();
    hsl.h = (hsl.h + 180) % 360;
    return TinyColor.fromHSL(hsl);
  }

  Color get color => _color;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TinyColor &&
          runtimeType == other.runtimeType &&
          color == other.color;

  @override
  int get hashCode => color.hashCode;

  @Deprecated('Use == instead.')
  bool equals(Object other) => this == other;
}

extension _ColorExtension on Color {
  Color clone() => Color.fromARGB(alpha, red, green, blue);
}
